packages/backend/server/src/modules/doc/history.ts

```diff
@@ -0,0 +1,237 @@
+    private readonly metrics: Metrics
+  ) {}
+
+  @OnEvent('doc:manager:snapshot:beforeUpdate')
+  async onDocUpdated(snapshot: Snapshot, forceCreate = false) {
+    const last = await this.last(snapshot.workspaceId, snapshot.id);
+
+    let shouldCreateHistory = false;
+
+    if (!last) {
+      // never created
+      shouldCreateHistory = true;
+    } else if (last.timestamp === snapshot.updatedAt) {
+      // no change
+      shouldCreateHistory = false;
+    } else if (
+      // force
+      forceCreate ||
+      // last history created before interval in configs
+      last.timestamp.getTime() <
+        snapshot.updatedAt.getTime() - this.config.doc.history.interval
+    ) {
+      shouldCreateHistory = true;
+    }
+
+    if (shouldCreateHistory) {
+      await this.db.snapshotHistory
+        .create({
+          select: {
+            timestamp: true,
+          },
+          data: {
+            workspaceId: snapshot.workspaceId,
+            id: snapshot.id,
+            timestamp: snapshot.updatedAt,
+            blob: snapshot.blob,
+            state: snapshot.state,
+            expiredAt: await this.getExpiredDateFromNow(snapshot.workspaceId),
+          },
+        })
+        .catch(() => {
+          // safe to ignore
+          // only happens when duplicated history record created in multi processes
+        });
+      this.metrics.docHistoryCounter(1, {});
+      this.logger.log(
+        `History created for ${snapshot.id} in workspace ${snapshot.workspaceId}.`
+      );
+    }
+  }
+
+  async list(
+    workspaceId: string,
+    id: string,
+    before: Date = new Date(),
+    take: number = 10
+  ) {
+    return this.db.snapshotHistory.findMany({
+      select: {
+        timestamp: true,
+      },
+      where: {
+        workspaceId,
+        id,
+        timestamp: {
+          lte: before,
+        },
+        // only include the ones has not expired
+        expiredAt: {
+          gt: new Date(),
+        },
+      },
+      orderBy: {
+        timestamp: 'desc',
+      },
+      take,
+    });
+  }
+
+  async count(workspaceId: string, id: string) {
+    return this.db.snapshotHistory.count({
+      where: {
+        workspaceId,
+        id,
+        expiredAt: {
+          gt: new Date(),
+        },
+      },
+    });
+  }
+
+  async get(workspaceId: string, id: string, timestamp: Date) {
+    return this.db.snapshotHistory.findUnique({
+      where: {
+        workspaceId_id_timestamp: {
+          workspaceId,
+          id,
+          timestamp,
+        },
+        expiredAt: {
+          gt: new Date(),
+        },
+      },
+    });
+  }
+
+  async last(workspaceId: string, id: string) {
+    return this.db.snapshotHistory.findFirst({
+      where: {
+        workspaceId,
+        id,
+      },
+      select: {
+        timestamp: true,
+      },
+      orderBy: {
+        timestamp: 'desc',
+      },
+    });
+  }
+
+  async recover(workspaceId: string, id: string, timestamp: Date) {
```

Comment:
I am thinking about the recover api in the history recovery workflow. I assume when calling `recover` there will not broadcast update events to peers right? So the workflow is:

1. ui: user chooses a snapshot in history view and click recover
2. ui: send "recover" api to the server
3. server: create a new snapshot and push to history
4. server: replace the latest snapshot with the given snapshot
5. ui: replace the given page's doc & apply the snapshot
6. ui: send updates to server
7. server: updates the snapshot
8. server: broadcast updates to other peers

It seems to me that there are some redundant in no.3/4 & 7. The issue is that when the user updates the local doc, it will send updates to the server to update the doc, but it might be already be updated.

**Updates:**
I searched yjs repo and seems this is [a discussed topic](https://github.com/yjs/yjs/issues/159#issuecomment-1525961593). It seems not possible force reset a doc to a certain state, unless create a new doc and replace the existing subodc in the doc tree with the new doc.

> Once a document that was created using createDocFromSnapshot syncs with other docs the discarded changes will be restored - unless you find a way to force everyone to discard the changes simultaneously.

Reply:
This feature has been implemented in the internal y-indexeddb before: https://github.com/toeverything/AFFiNE/blob/00c11d40cff9fa203f11864a01a927e52cbeaf82/packages/common/y-indexeddb/src/index.ts#L34-L72
We only need to implement it once in the server and then broadcast the changes

Reply:
oh I forget that, and the bad news is, the recover is doing in main server and the broadcast happens in sync server. we might need some hack

Reply:
Consider that we have stricter permission management requirements in the future (such as the team version), I don't think it is safe to roll back the version in frontend
Malicious users can compile the malicious version to roll back and broadcast it to the server
It is difficult for the server to prevent the broadcast of this kind of update information before we implement a update analysis service